austrakka/components/sequence/sync/sync_workflow.py

```diff
@@ -0,0 +1,474 @@
+
+    output_dir = get_output_dir(sync_state)
+    file_path = os.path.join(
+        output_dir,
+        sync_state[OBSOLETE_OBJECTS_FILE_KEY])
+
+    move_delete_targets_to_trash(file_path, output_dir, trash_dir_path)
+
+    # Delete any empty directories beneath output_dir.
+    remove_empty_dirs(output_dir, [TRASH_DIR])
+
+    # Remove the delete target file. It'll be recreated on the next run.
+    os.remove(file_path)
+
+    # Delete the intermediate manifest. It's no longer needed.
+    # It'll be recreated on the next run.
+    os.remove(os.path.join(
+        output_dir,
+        sync_state[INTERMEDIATE_MANIFEST_FILE_KEY]))
+
+    sync_state[CURRENT_STATE_KEY] = SName.DONE_PURGING
+    sync_state[CURRENT_ACTION_KEY] = Action.set_state_up_to_date
+    logger.success(f'Finished: {Action.purge}')
+
+
+def set_state_up_to_date(sync_state: dict):
+    sync_state[CURRENT_STATE_KEY] = SName.UP_TO_DATE
+    sync_state[CURRENT_ACTION_KEY] = Action.set_state_pulling_manifest
+
+
+def finalise_each_file(int_med, sync_state):
+    output_dir = get_output_dir(sync_state)
+    for index, _ in int_med.iterrows():
+
+        dest = os.path.join(
+            output_dir,
+            int_med.at[index, SAMPLE_NAME_KEY],
+            int_med.at[index, FILE_NAME_ON_DISK_KEY])
+
+        if int_med.at[index, STATUS_KEY] == DRIFTED:
+            src = os.path.join(
+                output_dir,
+                int_med.at[index, SAMPLE_NAME_KEY],
+                int_med.at[index, HOT_SWAP_NAME_KEY])
+
+            logger.info(f'Hot swapping: {dest}')
+            os.rename(src, dest)
+            logger.success(f'Done hot swapping: {dest}')
+
+            int_med.at[index, STATUS_KEY] = DONE
+
+        elif int_med.at[index, STATUS_KEY] == MATCH or\
+                int_med.at[index, STATUS_KEY] == DOWNLOADED:
+
+            int_med.at[index, STATUS_KEY] = DONE
+            logger.success(f'Done: {dest}')
+
+        else:
+            raise WorkflowError(
+                f'Reach an impossible state in the depths of {Action.finalise}. '
+                f'Expecting each file state to be only "{MATCH}", '
+                f'"{DOWNLOADED}", or "{DRIFTED}" but got something else. '
+                f'The caller, probably {Action.finalise}, should have checked '
+                f'my inputs before calling me.')
+
+    save_int_manifest(int_med, sync_state)
+
+
+def detect_and_record_obsolete_files(int_med, sync_state):
+    logger.info('Checking for obsolete files..')
+
+    # Get the list of files on disk. The array is a list of (full_path,
+    # file_name_only)
+    files_on_disk = []
+    obsoletes = pd.DataFrame({
+        FILE_PATH_KEY: [],
+        FILE_NAME_KEY: [],
+        DETECTION_DATE_KEY: []
+    })
+
+    for (
+            root_dir,
+            dir_names,
+            file_names) in os.walk(
+            get_output_dir(sync_state)):
+        dir_names[:] = [d for d in dir_names if d not in [TRASH_DIR]]
+        for name in file_names:
```

Comment:
I'd like to add a more paranoid check here and also report on ignored files - will suggest change via PR though